### PR TITLE
How to authenticate users in nestjs

### PR DESCRIPTION
**Pull request by gathoni-k**
#### Please note that in case of below checklist not updated accordingly, the maintainers have the right to immediately close your pull request.

- [ x ] You created a branch from `dev` branch.
- [ x ] You PR is raised on `dev` branch and not on`master`.
- [ x ] Blog follows [Blog Guidelines](https://github.com/LoginRadius/engineering-portal/blob/master/GUIDELINES.md).
- [ x ] You have read the [Contribution Guidelines](https://github.com/LoginRadius/engineering-portal/blob/master/CONTRIBUTING.md) before creating this PR.
- [ x ] Checked for any duplicate blog entry.
- [ x ] Checked for all grammatical errors in the blog.
- [ x ] You have not published the blog content anywhere else on the internet on platforms like medium, dev.to etc.
- [x ] The content is not copy-pasted from anywhere else on the Internet.

##### Description of change
A tutorial on how to authenticate users using the Authentication API provided by LoginRadius in NestJS.

### CODE REVIEW

**Comment by raghunath-r-a** [comment]
Hi @gathoni-k We're moving your PR to Phase - II.

For facilitating easier reference for readers, please add the sample/example code used in the following repo by creating a separate pull request: https://github.com/LoginRadius/engineering-blog-samples

**Comment by gathoni-k** [comment]
Hey!

Okay, I will.

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+})
+export class AppModule {}
+```
+
+Add the API key, API secret, and [Secure One Time Token(SOTT)](https://www.loginradius.com/docs/developer/concepts/sott/) to the.env file.
+
+```
+APP_NAME=<your app name>
+API_KEY=<your api key>
+API_SECRET=,your api secret>
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:90]: Kindly change the `,` to `<` for API_SECRET.

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+To create a user, you need to create a signup route that will accept the email and password.
+
+Since you are using TypeScript, define the DTO (Data Transfer Object) schema to validate the user data passed in the request body.
+
+In the `auth` folder, add the `dto` folder and create a `UserDTO` class in the `user.dto.ts` file.
+
+```JavaScript
+export class UserDto {
+  email: string,
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:121]: Remove the `,` here.

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    let response = await this.authService.signup(registerUserDto)
+    return response
+  }
+}
+```
+
+Before registering the user, validate if the email is already in use.
+
+Start by configuring loginradius-sdk.
+
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:191]: Before importing, add a line stating the command used to install `loginradius-sdk`, i.e `npm i loginradius-sdk`

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+## Access Protected Routes
+
+For protected routes, like accessing a user dashboard, the user will need to send the access token with the request. The access token will then be verified, and if valid, the application will be granted access.
+
+The user will need to store the `accessToken`. In this tutorial, you will be storing the token in the authorization header as a bearer token. Another alternative would be to use HTTP-only cookies.
+
+In NestJS, guards are responsible for handling authorization. They determine whether a request will be handled by the route.
+
+In `auth.guard.js`, add the following code.
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:326]: Change `auth.guard.js` to `auth.guard.ts`

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+    async canActivate(context: ExecutionContext): Promise<boolean> {
+        const request: Request = context.switchToHttp().getRequest();
+
+        // Extract the access token from the authorization header
+        const authheader = request.header('Authorization');
+        const token = authheader && authheader.split(" ")[1];
+        try {
+            let authorizedMsg = await this.authService.authenticate(token);
+            // Attach the authorized message to the request. You could also attach the user information.
+            request['isAuthorized'] = "Authorrized"
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:348]: Spelling of `Authorized` is wrong.

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+@Injectable()
+export class AuthService {
+    async signup(registerUserDto: UserDto) {
+      console.log('sign up')
+    }
+}
+```
+
+Note that you are also importing the user DTO and the loginradius-sdk at the top of the file.
+To execute the signup method in the signup route, inject it in the `auth.contoller.ts` file.
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:168]: `controller` is missplet in file name.

File: content/guest-post/nestjs-authentication-with-loginradius-api/index.md [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+          return error
+        }
+    }
+}
+```
+
+In the above code, you register a new user by passing in the user data to the authentication API. The `authUserRegistrationModel` object defines how the email and password will be stored in the database.
+
+## Log in the User
+
```
Comment (Akash-Patil) [inline content/guest-post/nestjs-authentication-with-loginradius-api/index.md:271]: In the Login Section, you can mention the code that needs to be injected in the `auth.controller.ts` file for the login method to work on login route, similar to as mentioned in the signup section.

**Review by Akash-Patil — CHANGES_REQUESTED** [review]
Hi @gathoni-k. Please note the following points: 

1. I have reviewed the blog and suggested a few changes. Kindly have a look. 
2. While testing the code, the `lrv2` variable is fetching/assigned an undefined value and hence the desired result is not returned on hitting the different routes. As suggested by @raghunath-r-a, please send a PR to https://github.com/LoginRadius/engineering-blog-samples with the demo code so that we can check for possible errors in code or if there is any other external issue from our end. After that, you can add a link to the demo in the blog.
3. Lastly, kindly rebase your branch to the latest commit of dev branch once the changes are made.

Regards,
@Akash-Patil

**Review by Akash-Patil — CHANGES_REQUESTED** [review]
Hi @gathoni-k, 

Changes look good. We have reviewed you demo code PR  and merged it. Now you update the blog links in the main blog and rebase your branch with dev branch so that we can merge the blog.

Regards 
@Akash-Patil

**Review by Akash-Patil — APPROVED** [review]

**Pull request merged**